src/kubernetes/api_watcher.rs

```diff
@@ -0,0 +1,160 @@
+//! A watcher based on the k8s API.
+
+use super::{
+    client::Client,
+    stream as k8s_stream,
+    watch_request_builder::WatchRequestBuilder,
+    watcher::{self, Watcher},
+};
+use futures::{
+    future::BoxFuture,
+    stream::{BoxStream, Stream},
+};
+use http02::StatusCode;
+use hyper13::Error as BodyError;
+use k8s_openapi::{WatchOptional, WatchResponse};
+use snafu::{ResultExt, Snafu};
+
+/// A simple watcher atop of the Kubernetes API [`Client`].
+pub struct ApiWatcher<B>
+where
+    B: 'static,
+{
+    client: Client,
+    request_builder: B,
+}
+
+impl<B> ApiWatcher<B>
+where
+    B: 'static,
+{
+    /// Create a new [`ApiWatcher`].
+    pub fn new(client: Client, request_builder: B) -> Self {
+        Self {
+            client,
+            request_builder,
+        }
+    }
+}
+
+impl<B> ApiWatcher<B>
+where
+    B: 'static + WatchRequestBuilder,
+    <B as WatchRequestBuilder>::Object: Send + Unpin,
+{
+    async fn invoke(
+        &mut self,
+        watch_optional: WatchOptional<'_>,
+    ) -> Result<
+        impl Stream<
+                Item = Result<
+                    WatchResponse<<B as WatchRequestBuilder>::Object>,
+                    k8s_stream::Error<BodyError>,
+                >,
+            > + 'static,
+        watcher::invocation::Error<invocation::Error>,
+    > {
+        // Prepare request.
+        let request = self
+            .request_builder
+            .build(watch_optional)
+            .context(invocation::RequestPreparation)?;
+        trace!(message = "request prepared", ?request);
+
+        // Send request, get response.
+        let response = self
+            .client
+            .send(request)
+            .await
+            .context(invocation::Request)?;
+        trace!(message = "got response", ?response);
+
+        // Handle response status code.
+        let status = response.status();
+        if status != StatusCode::OK {
+            let source = invocation::Error::BadStatus { status };
+            let err = if status == StatusCode::GONE {
+                watcher::invocation::Error::Desync { source }
+            } else {
+                watcher::invocation::Error::Other { source }
+            };
+            Err(err)?;
+        }
+
+        // Stream response body.
+        let body = response.into_body();
+        Ok(k8s_stream::body(body))
+    }
+
+    async fn invoke_boxed_stream(
```

Comment:
This function feels like so much boilerplate for a simple thing. :(

Reply:
Yeah, I tried to rewrite the code without it, but ended up with even more code. Maybe I'm doing it wrong, idk.

Reply:
Ok :) No problem.